### PR TITLE
Make it possible to build Domoticz when cloned from a GIT repo

### PR DESCRIPTION
**Pull request by ikkemaniac**

### CODE REVIEW

**Comment by gizmocuz** [comment]
I do care about revision numbers as this is used for the internal version display
one way could be:
git rev-list --count HEAD

**Comment by ikkemaniac** [comment]
I don't quite understand your comment? But why would you want to use the count number?

I'm using git rev-parse --short HEAD. This will give you the clean, short git hash domoticz was build from when using git instead of svn. The hash is familiar and easy to understand for git users instead of the count number you describe. Futhermore this will also prevent confusion about talking svn revision numbers and users using git. The version shown in the GUI is in this format: 'V2.g9d15b6e'

**Comment by gizmocuz** [comment]
Because we want to show a clean version number, also when end-users 
report to have found a problem in version xxxx, then we know what to 
look for
yes, did you see your version number, looks kinda strange indeed

part your correct of course... maybe at startup we should show the hash 
as well.

this is i think the only thing i do not like about git

But i still am going to transfer to it... it will take some more days as 
the makebeta/makerelease and upload also needs to be changed

and i miss a place where i can upload archives from the build scripts, 
it seems the github pages are not accessible with scp ?

**Comment by ikkemaniac** [comment]
> Because we want to show a clean version number, also when end-users report to have found a problem in version xxxx, then we know what to look for 
> yes, did you see your version number, looks kinda strange indeed part your correct of course... maybe at startup we should show the hash as well. this is i think the only thing i do not like about git 

I understand what you are saying, git revision numbers are hex hashses. But this is the power of GIT! every commit in a repo is unique. When some shows the git hash you know exactly what revision they are on. I dis advice highly to use the git count number, it shouldn't be used for versioning at all! The count number is not unique identifiable and can be differen per user/checkout repo. See the man pages (http://git-scm.com/docs/git-rev-list):
```
--count

    Print a number stating how many commits would have been listed, and suppress all other output. When used together with --left-right, instead print the counts for left and right commits, separated by a tab. When used together with --cherry-mark, omit patch equivalent commits from these counts and print the count for equivalent commits separated by a tab.

``` 

> But i still am going to transfer to it... it will take some more days as the makebeta/makerelease and upload also needs to be changed and i miss a place where i can upload archives from the build scripts, it seems the github pages are not accessible with scp ?

Please don't implement it at all if you don't plan on using git hashes.
If you do plan on using it I'm happy to help update makebeta/makrelease.

cheers

**Comment by gizmocuz** [comment]
Okey, seems your good with GIT (and i am not, but just made my first commit test, and it worked)
....

There are many things todo:

- we need an incrementing version number... 
Can we install a hook on github that increments the number on every commit/merge?

makebeta/makerelease
- it should check the current (hash?) against the online hash, to see if it needs to pull the new code.
- it should get the above incrementing number, and use this as the version for the web display, and this will also be used in the webupdate to check if there is a new version
- it should upload the archive to github, if not possible we stay on sourceforge for the release archives

CMakefile
- indeed drop the SVN stuff, but also use the above incremental version number

Application:
- we still need to display this unique version number, but maybe with a tooltip over it, it should display also the hash
- at startup we can also display the hash (V2.2847.jkdhsfg)

is this an idea? its still a lot of work todo ;)

**Comment by ikkemaniac** [comment]
All your ideas are good. I just shows me how much stuff I didn't think about.

Everything you say is possible with some minor tweaking. I'm happy to help with this if you're ok with it. The main question you need to as yourself is where you want to keep your main repo: svn of git? 
If you're not that familiar with git i would keep it on svn. This would mean git FOLLOWS svn (like it is now). I see that you are not using  the 'git svn' tool. How are keeping the git repo in sync with your svn repo? 

I've noticed that you've completely dropped the master branch on github and pushed a completely new branch as master. In the git kingdom this is a very big no,no! My git repo was no longer connected (forked) from yours and made it PITA to get it in sync with you again. Because of this you've also lost all other people that originally forked your repo.

Releasing will be more complicated if you want the include GIT because of versioning. It would better if you start using the version numbering scheme in the CMakeLists.txt file in order to simplify versioning between svn and git. Right now this is still on 'MAJOR'=1 and 'MINOR'=0 while the GUI shows 'V2'.

The second thing is that you need to make git follow svn. So while it's possible to make a git hook to increment the 'development' revision number i think you shouldn't. Because if you do, you need to put this number some back in you svn repo. The best would be if you would just commit changes from git in svn and than push your new svn commits to the git master branch (like you are doing now).

So in short, it's possible, I'm willing to help; just think about your work methods on forehand. Otherwise it's going to be a PITA.

cheers

**Comment by gizmocuz** [comment]
the idea is to move 100% to github.
At the moment i am copying the changed files in a git repo and do a commit (all by hand)

**Comment by ikkemaniac** [comment]
Ok good to know.

I think you'll love git once you're used to it.

In my opinion the best thing to do, if you're fully commited to using git, would be to drop the whole 4 digit patch versioning. In true git style this version number should be replaced by the (short, 7 chars) git hash. 

Each and every commit is uniquely identifiable by its hash (which is actually 40 chars long). So commits can be easily distinguished from another, even when on different branches.

For debugging it would pose no problem at all if you'll make the GUI version show the git hash. The git hash is easily recognized as a hash, it's unique (you can even see if someone is using a commit that is not from your repo but their own work since it just doesn't exist in you repo).

My preferred branching model for the Domoticz repo would be as described by [url=http://nvie.com/posts/a-successful-git-branching-model/]this article[/url]
- master branch, this the 'stable' branch where the releases live, including proper tags.
- develop branch, this is the 'beta release' or 'nightly builds' branch. This would be the branch that you are pushing you development to and test it.
- if you want to support pre-compiled images i would create a separate folder in the repo, for example called 'images'. This folder would containg only a single file in the master and develop branch stating: "Pre-compiled images are stored in the images branch!"
A dedicated images branch would be created. In this branch you would then put the actualy image files you want to make available into the images folder. This can be done by simple making a new commit to branch with the correct files. (the "downside" is that older versions will always stay available due the availability of the git history of the branch! If this might be a problem you can override normal git behaviour and override the current commit in that branch.
- all kinds of different working branches. These are the branches you'll actually commit all you work to while developing or bugfixing.

pls take a look at the link, they have a very nice graphic presentation about this 'branching'.

So the biggest hurdle would be that you need to get used to hex version numbers :)

I'll take a look at the makebeta/makerelease files and propose a change.

**Comment by ikkemaniac** [comment]
see pullrequest #3 for a working makebeta.git.sh script.

**Comment by gizmocuz** [comment]
Thanks a lot! I will check it out

**Comment by ikkemaniac** [comment]
For now this script only looks for the status of the remote branch. It could happen that your own branch is ahead of the remote which is not detected.
